### PR TITLE
feat: add transformer to convert tags to structured properties

### PR DESCRIPTION
**Pull request by sgomezvillamor**
## Problem

Users need to convert tags into structured properties to leverage the more powerful structured properties system. Currently:
- Snowflake source has this capability but it's source-specific
- Other sources (Tableau, dbt, Kafka, etc.) emit tags but have no way to convert them
- Users are forced to write custom post-processing scripts
- Tags are simple strings; structured properties provide typing, validation, and semantic meaning

## Solution

This PR adds a **generic transformer** that works with any source emitting tags:

1. **New `MultipleAspectTransformer` framework pattern**
   - Extends base transformer to emit multiple aspects on-the-fly (vs. accumulating state until end-of-stream)
   - More efficient for large ingestion runs

2. **`TagsToStructuredPropertiesTransformer`** with two modes:
   - **Key-value tags**: Parse tags like `dept:Finance` into structured properties
   - **Keyword mapping**: Map simple tags to properties via configuration (inverted index)

3. **Configurable behavior**:
   - Keep or remove original tags
   - Support for datasets, charts, dashboards, dataJobs, dataFlows, schema fields
   - Flexible property prefix configuration
   - Warning logs for unmatched tags (non-blocking)

## Usage Example

```yaml
transformers:
  - type: tags_to_structured_properties
    config:
      # Key-value mode
      process_key_value_tags: true
      key_value_property_prefix: "io.company."
      
      # Keyword mapping mode
      tag_structured_property_map:
        "io.company.department":
          values: ["Finance", "Sales", "Marketing"]
        "io.company.sensitivity":
          values: ["PII", "Confidential"]
      
      remove_original_tags: false
```

### Ingestion will fail if:
- **Invalid property value**: Tag value doesn't match allowed values defined in the structured property
- **Property doesn't exist**: Transformed property URN references a non-existent structured property

These failures trigger backend validation errors and are **expected behavior** - they indicate configuration issues that require either:
- Updating the transformer config (`tag_structured_property_map` or `key_value_property_prefix`)
- Updating the structured property definition (allowed values, cardinality, etc.)

### Silently ignored (debug log only):
- **Unmatched tags**: Tags that don't match any mapping configuration are logged but don't block ingestion
- This allows partial transformation - only configured tags are converted

## Testing

- 21 unit tests covering both transformer modes and edge cases
- Tests for base transformer framework additions
- All existing tests passing
- End-to-end testing validated with Tableau source

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        if self.config.remove_original_tags:
+            output_aspects.append(("globalTags", None))
+        else:
+            output_aspects.append(("globalTags", cast(Optional[Aspect], tags_aspect)))
+
+        # Second: structured properties (if any were mapped)
+        if properties_dict:
+            structured_props = StructuredPropertiesClass(
+                properties=[
+                    StructuredPropertyValueAssignmentClass(
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py:170]: This will only work if the property already exists. If it an expected limitation the we should document it.

Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py:170]: Yes, I did that on purpose, by design: The transformer is not creating struct props dynamically. Instead, struct props must be created.

It is documented already:

```
- Structured properties must be created in DataHub before running the transformation
```

Also, in the PR it is mentioned that:

```
Ingestion will fail if:
- Invalid property value: Tag value doesn't match allowed values defined in the structured property
- Property doesn't exist: Transformed property URN references a non-existent structured property
```

I have added commit f2ed23977871074e3d9236bd4a8bc1b7108e5ca3 to emphasize this in the docs

File: metadata-ingestion/src/datahub/ingestion/transformer/base_transformer.py [hunk 2]
```diff
@@ -198,11 +272,165 @@ def _transform_or_record_mcpw(
...
+
+                    mcp: MetadataChangeProposalWrapper = MetadataChangeProposalWrapper(
+                        entityUrn=urn,
+                        entityType=structured_urn.get_type(),
+                        systemMetadata=(
+                            last_seen_mcp.systemMetadata
+                            if last_seen_mcp
+                            else last_seen_mce_system_metadata
+                        ),
+                        aspectName=self.aspect_name(),  # type: ignore
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/transformer/base_transformer.py]: Why do we need to ignore types?

File: metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+        Returns tuples of (aspect_name, aspect):
+        - First tuple: the original globalTags (possibly removed if config says so)
+        - Second tuple: the new structuredProperties aspect (if any properties were mapped)
+        """
+
+        if not aspect:
+            return []
+
+        tags_aspect = cast(GlobalTagsClass, aspect)
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py]: do we need to check if aspect is GlobalTagsClass or not?

Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/transformer/tags_to_structured_properties.py]: isInstance would be more appropriate here

File: metadata-ingestion/docs/transformer/dataset_transformer.md [hunk 4]
```diff
@@ -809,6 +809,115 @@ transformers:
...
+
+### Config Details
+
+| Field                         | Required | Type                                        | Default | Description                                                               |
+| ----------------------------- | -------- | ------------------------------------------- | ------- | ------------------------------------------------------------------------- |
+| `process_key_value_tags`      |          | boolean                                     | `false` | Enable parsing of key:value formatted tags                                |
+| `key_value_separator`         |          | string                                      | `:`     | Separator character for key-value tags                                    |
+| `key_value_property_prefix`   |          | string                                      | `""`    | Prefix to add to property IDs derived from key-value tags                 |
+| `tag_structured_property_map` |          | map[property_id, StructuredPropertyMapping] | `{}`    | Map of structured property IDs to tag keyword configurations              |
```
Comment (treff7es) [inline metadata-ingestion/docs/transformer/dataset_transformer.md:832]: This should be a set, shouldn't it?

File: metadata-ingestion/src/datahub/ingestion/transformer/base_transformer.py (hunk 2 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/transformer/base_transformer.py]: addressed in commit 974ffb52b338c6b17222bd7231c26239b4c01b93

File: metadata-ingestion/docs/transformer/dataset_transformer.md (hunk 4 above)
Comment (sgomezvillamor) [reply metadata-ingestion/docs/transformer/dataset_transformer.md:832]: not, it's a map (or a set of key-value pairs 😅 )

eg

```
              tag_structured_property_map:
                "io.company.department":
                  values: ["Finance", "Sales", "Marketing"]
                "io.company.dataClassification":
                  values: ["PII", "Confidential"]
```
